### PR TITLE
Add list of available callbacks in each component

### PR DESCRIPTION
**Pull request by varsill**

### CODE REVIEW

File: lib/membrane/docs_helper.ex [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        {:callback, {:"::", _line1, [{name, _line2, array} | _rest]}, _subtree} ->
+          {{name, if(is_nil(array), do: 0, else: length(array))}, module}
+
+        {:callback,
+         {:when, _line1,
+          [
+            {:"::", _line2,
+             [
+               {name, _line3, array},
+               :ok
```
Comment (mat-hek) [inline lib/membrane/docs_helper.ex:63]: This seems weird, are you sure it's correct? Maybe it's possible to remove `when` when present and try parsing again with the same code

File: lib/membrane/docs_helper.ex [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+      docstring <>
+        """
+        ## List of available callbacks
+        #{generate_docstring_with_list_of_callbacks(module, inherited_modules_list, starting_with)}
+        """
+
+    Module.put_attribute(module, :moduledoc, {line, new_docstring})
+  end
+
+  defp generate_docstring_with_list_of_callbacks(module, modules_list, starting_with) do
```
Comment (mat-hek) [inline lib/membrane/docs_helper.ex:28]: grouping callback by behaviours and explaining why we include them would be great

**Review by mat-hek — CHANGES_REQUESTED** [review]

File: lib/membrane/bin.ex [hunk 3]
```diff
@@ -342,4 +342,6 @@ defmodule Membrane.Bin do
                      handle_parent_notification: 3
     end
   end
+
+  Membrane.DocsHelper.add_callbacks_list_to_moduledoc(__MODULE__, [], "handle_")
```
[the remark this answers is not in the document]
Comment (varsill) [reply lib/membrane/bin.ex:346]: done

File: lib/membrane/docs_helper.ex [hunk 4]
```diff
@@ -0,0 +1,71 @@
+defmodule Membrane.DocsHelper do
```
[the remark this answers is not in the document]
Comment (varsill) [reply lib/membrane/docs_helper.ex:1]: indeed, done

File: lib/membrane/docs_helper.ex (hunk 1 above)
Comment (varsill) [reply lib/membrane/docs_helper.ex:63]: well, it works fine, but I will implement your suggestion

Comment (varsill) [reply lib/membrane/docs_helper.ex:63]: done

File: lib/membrane/core/docs_helper.ex [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+          {parse_callback_ast(callback_ast), module}
+
+        {:callback, callback_ast, _subtree} ->
+          {parse_callback_ast(callback_ast), module}
+      end
+    )
+  end
+
+  defp parse_callback_ast({:"::", _line1, [{name, _line2, array} | _rest]}) do
+    {name, if(is_nil(array), do: 0, else: length(array))}
```
Comment (mat-hek) [inline lib/membrane/core/docs_helper.ex:71]:
```suggestion
  defp parse_callback_ast({:"::", _line1, [{name, _line2, args} | _rest]}) do
    {name, if(is_nil(args), do: 0, else: length(args))}
```

File: lib/membrane/core/docs_helper.ex [hunk 6]
```diff
@@ -0,0 +1,73 @@
+defmodule Membrane.Core.DocsHelper do
+  @moduledoc """
```
Comment (mat-hek) [inline lib/membrane/core/docs_helper.ex:2]:
```suggestion
  @moduledoc false
```

File: lib/membrane/core/docs_helper.ex [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+    new_docstring =
+      docstring <>
+        """
+        ## List of available callbacks
+        Below there is a list of all the callbacks available in a module, that implements `#{module}` behaviour.
+        We have put it for your convenience, as some of these callbacks are not directly defined in that module and
+        their specification is available in different modules.
+
+        The callbacks available in `#{module}` behaviour:
```
Comment (mat-hek) [inline lib/membrane/core/docs_helper.ex:26]:
```suggestion
        Below there is a list of all the callbacks available in a module, that implements `#{inspect(module)}` behaviour.
        We have put it for your convenience, as some of these callbacks are not directly defined in that module and
        their specification is available in different modules.

        The callbacks available in `#{inspect(module)}` behaviour:
```

**Review by mat-hek — CHANGES_REQUESTED** [review]

File: lib/membrane/docs_helper.ex (hunk 2 above)
Comment (varsill) [reply lib/membrane/docs_helper.ex]: done

File: lib/membrane/core/docs_helper.ex (hunk 7 above)
Comment (varsill) [reply lib/membrane/core/docs_helper.ex]: done

**Review by mat-hek — CHANGES_REQUESTED** [review]
+ grouping by behaviour seems still missing

File: lib/membrane/core/docs_helper.ex [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+    new_docstring =
+      docstring <>
+        """
+        ## List of available callbacks
+        Below there is a list of all the callbacks available in a module, that implements `#{inspect(module)}` behaviour.
+        We have put it for your convenience, as some of these callbacks might not be directly defined in that module and
+        their specification is available in different modules.
+
+        The callbacks available in `#{module}` behaviour:
```
[the remark this answers is not in the document]
Comment (varsill) [reply lib/membrane/core/docs_helper.ex:24]: ups, I have forgotten about checking if that changes is applicable in other places :D 
done

File: lib/membrane/bin.ex [hunk 9]
```diff
@@ -342,4 +342,6 @@ defmodule Membrane.Bin do
                      handle_parent_notification: 3
     end
   end
+
+  DocsHelper.add_callbacks_list_to_moduledoc(__MODULE__)
```
Comment (mat-hek) [inline lib/membrane/bin.ex:346]:
```suggestion
```

**Review by mat-hek — CHANGES_REQUESTED** [review]

File: lib/membrane/core/docs_helper.ex [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+        "`c:#{inspect(module)}.#{Atom.to_string(name)}/#{arity}`"
+      end)
+
+    """
+    #{"* " <> Enum.join(callbacks_names, "\n* ")}
+    """
+  end
+
+  defp get_callbacks_in_compiled_module(module) do
+    Enum.map(module.behaviour_info(:callbacks), &{&1, module})
```
Comment (mat-hek) [inline lib/membrane/core/docs_helper.ex:60]:
```suggestion
    module.behaviour_info(:callbacks) |> Enum.sort() |> Enum.map(&{&1, module})
```

**Review by mat-hek — CHANGES_REQUESTED** [review]

File: lib/membrane/core/docs_helper.ex [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+    #{Enum.zip(all_modules, callbacks_per_module_list) |> Enum.map_join(fn {module, callbacks_list} -> """
+      `#{inspect(module)}`
+      #{generate_docstring_with_list_of_callbacks(callbacks_list)}
+      """ end)}
+    """
+
+    Module.put_attribute(module, :moduledoc, {line, new_docstring})
+  end
+
+  defp generate_docstring_with_list_of_callbacks(list_of_callbacks) do
```
Comment (mat-hek) [inline lib/membrane/core/docs_helper.ex:47]: the module could be passed to this function instead of being put in a tuple with each callback's name

**Review by mat-hek — CHANGES_REQUESTED** [review]

File: lib/membrane/bin.ex (hunk 9 above)
Comment (varsill) [reply lib/membrane/bin.ex:346]: done

File: lib/membrane/pipeline.ex [hunk 12]
```diff
@@ -494,4 +495,6 @@ defmodule Membrane.Pipeline do
                      handle_call: 3
     end
   end
+
+  DocsHelper.add_callbacks_list_to_moduledoc(__MODULE__)
```
[the remark this answers is not in the document]
Comment (varsill) [reply lib/membrane/pipeline.ex:499]: done

File: lib/membrane/core/docs_helper.ex (hunk 11 above)
Comment (varsill) [reply lib/membrane/core/docs_helper.ex]: done

File: lib/membrane/core/docs_helper.ex (hunk 10 above)
Comment (varsill) [reply lib/membrane/core/docs_helper.ex]: done

**Review by mat-hek — APPROVED** [review]

**Pull request merged**